### PR TITLE
Print autodiff types in the generated SIL

### LINKED ISSUE #63234: AutoDiff produces SIL that cannot be read back
This is a spin-off from https://github.com/apple/swift/issues/63207 to track the underlying issue. Currently autodiff produces SIL that cannot be parsed back. The reason is internal types that are never serialized as they are created in the artificial module.

Example SIL file is attached. Trying to parse it yields:
```
playground.sil:108:17: error: cannot find type '_AD__$s10playground5testAyyF1aL_5input10PlayGround1FVyS2dGAE1SVyS2dG_tF_bb0__PB__src_0_wrt_0' in scope
  %14 = struct $_AD__$s10playground5testAyyF1aL_5input10PlayGround1FVyS2dGAE1SVyS2dG_tF_bb0__PB__src_0_wrt_0 (%10 : $@callee_guaranteed (@inout F<Double, Double>) -> S<Double, Double>) // user: %16
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
playground.sil:117:136: error: cannot find type '_AD__$s10playground5testAyyF1aL_5input10PlayGround1FVyS2dGAE1SVyS2dG_tF_bb0__PB__src_0_wrt_0' in scope
sil private @$s10playground5testAyyF1aL_5input10PlayGround1FVyS2dGAE1SVyS2dG_tFTJpSpSr : $@convention(thin) (F<Double, Double>, @owned _AD__$s10playground5testAyyF1aL_5input10PlayGround1FVyS2dGAE1SVyS2dG_tF_bb0__PB__src_0_wrt_0) -> S<Double, Double> {
                                                                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

[playground.sil.txt](https://github.com/apple/swift/files/10509937/playground.sil.txt)

Tagging @BradLarson @rxwei @dan-zheng

### PR DESCRIPTION
**Pull request by asl**
The solution is a bit hacky as the types are created in the synthesised module, so we essentially need to pull all top-level decls, get all imported ones and match types by name. The types produced are also not quite correct as they are non-resilient and internal. Make them `@usableFromInline` to allow being `@frozen@.

Also, fixing SIL parsing issue while there.

Fixes #63234

### CODE REVIEW

**Comment by asl** [comment]
Tagging @BradLarson

**Comment by asl** [comment]
@swift-ci please test

**Review by dan-zheng — APPROVED** [review]
Nice!

(Probably good to get approval from @rxwei as the primary active code owner)

File: lib/SILOptimizer/Differentiation/LinearMapInfo.cpp
Comment (rxwei) [inline lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:137]: I think this should also respect the original linkage like line 90. We don't want to accidentally expose branching trace enums as ABI.

Comment (asl) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:137]: Well. Here is the problem. They are explicitly non-resilient due to `@frozen`. Therefore they should be either public or `@usableFromInline@`.

Comment (rxwei) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:137]: Not sure I understand. For a private or internal differentiable function, why should its derivative's branching trace enum be public or `@usableFromInline`?

Comment (asl) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:137]: Why they are `@frozen` then? This attributes essentially implies that the type is public-ish. SIL verifier errors in case of non-public frozen type.

Comment (rxwei) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:137]: > This attributes essentially implies that the type is public-ish.

Well, that's backwards...  It's a _bug_ that it's adding `@frozen` to non-ABI-public types.  The correct fix here would be to _not_ emit a `@frozen` attribute when it's not ABI-public.  Branching trace enums and linear map should have the exact same visibility and attributes.

Comment (asl) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:137]: Fair enough. I though it was intentional due to comments and how the code was organized and thought there are some corner cases that I'm not aware of.

Will be fixed in the subsequent "tuples" patch :)

**Pull request closed without merging**
